File: frontend/lib/api/phonebook.ts
import { http } from "./http";
import type { BroadcastRecipient, CsvImportResult } from "./messaging";

export interface PhoneBook {
  id: string;
  name: string;
  description: string | null;
  color: string | null;
  contact_count: number;
  created_at: string;
  updated_at: string;
}

export interface PhoneBookContact {
  id: string;
  phone: string;
  name: string | null;
  source: "csv" | "lead" | "manual";
  created_at: string;
  also_in_other_books: boolean;
}

export interface PhoneBookDetail extends PhoneBook {
  contacts: PhoneBookContact[];
}

export interface PhoneBookAddResult {
  added: number;
  skipped_duplicate: number;
  skipped_invalid: number;
  duplicate_phones: string[];
  cross_book_phones: string[];
}

export interface DuplicateAcrossBooks {
  phone: string;
  name: string | null;
  books: string[];
}

export const phoneBookApi = {
  list: () => http.get<{ items: PhoneBook[] }>("/phone-books"),
  get: (id: string) => http.get<PhoneBookDetail>(`/phone-books/${id}`),
  create: (body: { name: string; description?: string; color?: string }) =>
    http.post<PhoneBook>("/phone-books", body),
  update: (id: string, body: { name?: string; description?: string; color?: string }) =>
    http.patch<PhoneBookDetail>(`/phone-books/${id}`, body),
  remove: (id: string) => http.del<void>(`/phone-books/${id}`),
  addContacts: (id: string, body: { contacts?: BroadcastRecipient[]; lead_ids?: string[] }) =>
    http.post<PhoneBookAddResult>(`/phone-books/${id}/contacts`, body),
  removeContact: (id: string, contactId: string) =>
    http.del<void>(`/phone-books/${id}/contacts/${contactId}`),
  duplicates: () => http.get<{ items: DuplicateAcrossBooks[] }>("/phone-books/duplicates"),
  parseCsv: (file: File): Promise<CsvImportResult> => {
    const fd = new FormData();
    fd.append("file", file);
    return http.upload<CsvImportResult>("/phone-books/parse-csv", fd);
  },
};

File: backend/app/schemas/phone_book.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field


class PhoneBookCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    description: Optional[str] = None
    color: Optional[str] = Field(None, max_length=16)


class PhoneBookUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    description: Optional[str] = None
    color: Optional[str] = Field(None, max_length=16)


class PhoneBookContactIn(BaseModel):
    phone: str
    name: Optional[str] = None
    source: Optional[str] = "manual"


class PhoneBookAddContacts(BaseModel):
    contacts: List[PhoneBookContactIn] = Field(default_factory=list)
    lead_ids: List[str] = Field(default_factory=list)


class PhoneBookContactRow(BaseModel):
    id: str
    phone: str
    name: Optional[str] = None
    source: str
    created_at: datetime
    # True when this exact number also sits in at least one other phone book.
    also_in_other_books: bool = False

    model_config = {"from_attributes": True}


class PhoneBookResponse(BaseModel):
    id: str
    name: str
    description: Optional[str] = None
    color: Optional[str] = None
    contact_count: int = 0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PhoneBookDetail(PhoneBookResponse):
    contacts: List[PhoneBookContactRow] = []


class PhoneBookList(BaseModel):
    items: List[PhoneBookResponse]


class PhoneBookAddResult(BaseModel):
    added: int
    skipped_duplicate: int          # already in THIS phone book
    skipped_invalid: int            # not a usable phone number
    duplicate_phones: List[str] = []  # the ones skipped as duplicates
    cross_book_phones: List[str] = []  # added, but ALSO present in another book — flag for the user


class DuplicateAcrossBooks(BaseModel):
    phone: str
    name: Optional[str] = None
    books: List[str]  # phone book names this number appears in


class PhoneBookDuplicatesReport(BaseModel):
    items: List[DuplicateAcrossBooks]
